sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -757,7 +758,10 @@ private[sql] class SQLConf extends Serializable with CatalystConf with Logging {
 
   def variableSubstituteDepth: Int = getConf(VARIABLE_SUBSTITUTE_DEPTH)
 
-  def warehousePath: String = new Path(getConf(WAREHOUSE_PATH)).toString
+  def warehousePath: String = {
+    val path = new Path(getConf(WAREHOUSE_PATH))
+    FileSystem.get(path.toUri, new Configuration()).makeQualified(path).toString
```

Comment:
Hm, yes I can see how that would be tricky, mostly because of spaces in a path.
OK, how about this: if it parses as a URI with `java.net.URI`, parse it as a URI. If not, try to parse it as a URI with `java.io.File`. Does that present any problems I'm missing?

I think we have to support URIs because they always have been (?), and that also enables non-file URIs, which is implicitly supported right now. Although I'm reluctant to expand the possibilities here I tend to be persuaded that it's worth trying here specially.